torchao/prototype/common/bitpacking.py

```diff
@@ -1,101 +1,139 @@
-    data_size: int - the size of the small dtype in bits.
+    data: - a tensor of packed elements
+    element_dtype: - the dtype of the elements to unpack
     
     optional:
-    by_rows: bool - specifies whether to unpack... 
-        by rows: tensor(n,m) -> tensor(n*scale, m) 
-        or by columns: tensor(n,m) -> tensor(n,m*scale)
-        
-    defaults to rows because quantization is typically done by rows 
-    but choose the version which matches how you quantize as this improves memory accesses/performance
+    dimension: - the dimension to unpack along
+    
     
     Returns: torch.Tensor - a tensor of the unpacked elements.
     """
-    if by_rows:
-        return _unpack_by_rows(data, data_size, device)
-    else:
-        return _unpack_by_cols(data, data_size)
+    container_size = dtype_to_bits(data.dtype)
+    element_size = dtype_to_bits(element_dtype)
+    scale = container_size // element_size
     
-def pack(data, container_size, data_size, by_rows = True, device="cuda"):
+    unpacked = _unpack(data, element_size, container_size, scale, dimension, device)
+    
+    if element_dtype == "trinary":
+        unpacked = unpacked.to(torch.int8) - 1
+    return unpacked
+
+def _unpack(data, element_size, container_size, scale ,dim, device):
+    shape = data.shape
+    unpacked_data = torch.zeros(mod_shape(shape, shape[dim]*scale, dim), dtype=data.dtype).to(device)
+    nbits = (1 << element_size) - 1 # mask for the last dtype_size bits
+    for i in range(scale):
+        shift_amt = container_size - element_size * (i + 1)
+        slices = [slice(None)] * unpacked_data.ndim
+        slices[dim] = slice(i, None, scale)
+        unpacked_data[slices] = ((data >> shift_amt) & (nbits)).to(data.dtype)
+
+    # stack the unpacked data and reshape to the original shape
+    return unpacked_data.view(mod_shape(shape,scale*shape[dim], dim)) 
+    
+
+def pack(data: torch.Tensor,
+         element_dtype: Union[torch.dtype, str], # accepting strings for trinary until thats added to torch
+         dimension: Optional[int] = 0,
+         container_dtype: Optional[torch.dtype] = None,
+         device: Optional[str] = "cuda") -> torch.Tensor:
     """
-    Packs small dtype elements into a larger dtype.
-    Pads rows to be divisible by the scale.
+    Packs small dtype elements into a container of a larger dtype.
+    **Pads rows to be divisible by the scale**
+    TODO: support something like packing 8 uint 3s into 3 uint8s
     
     Inputs:
-    data: torch.Tensor - a tensor of unpacked elements of a small dtype.
-    container_size: int - the size of the large dtype in bits.
-    data_size: int - the size of the small dtype in bits.
+    data: a tensor of unpacked elements of a small dtype. The dtype used for the data will be used for the container.
+    dimension: the dimension to pack along
+    element_dtype: the dtype of the elements to pack
```

Comment:
I think keeping `element_size` (or maybe change to `element_bit_width`) is better. Then `dtype_to_bits()` doesn't need to support uint2-7 (so you can use `torch.finfo()` and `torch.iinfo()`)